subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -0,0 +1,417 @@
+static struct k_delayed_work retransmit_work;
+
+#define NUM_PENDINGS	CONFIG_LWM2M_ENGINE_MAX_PENDING
+#define NUM_REPLIES	CONFIG_LWM2M_ENGINE_MAX_REPLIES
+static struct zoap_pending pendings[NUM_PENDINGS];
+static struct zoap_reply replies[NUM_REPLIES];
+static struct zoap_block_context firmware_block_ctx;
+
+static void
+firmware_udp_receive(struct net_context *ctx, struct net_pkt *pkt, int status,
+		     void *user_data)
+{
+	struct zoap_pending *pending;
+	struct zoap_reply *reply;
+	struct zoap_packet response;
+	int header_len, r;
+	const u8_t *token;
+	u8_t tkl;
+	struct sockaddr from_addr;
+
+	/* Save the from address */
+#if defined(CONFIG_NET_IPV6)
+	if (net_pkt_family(pkt) == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&from_addr)->sin6_addr,
+				&NET_IPV6_HDR(pkt)->src);
+		net_sin6(&from_addr)->sin6_port = NET_TCP_HDR(pkt)->src_port;
+		net_sin6(&from_addr)->sin6_family = AF_INET6;
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+	if (net_pkt_family(pkt) == AF_INET) {
+		net_ipaddr_copy(&net_sin(&from_addr)->sin_addr,
+				&NET_IPV4_HDR(pkt)->src);
+		net_sin(&from_addr)->sin_port = NET_TCP_HDR(pkt)->src_port;
+		net_sin(&from_addr)->sin_family = AF_INET;
+	}
+#endif
+
+	/*
+	 * zoap expects that buffer->data starts at the
+	 * beginning of the CoAP header
+	 */
+	header_len = net_pkt_appdata(pkt) - pkt->frags->data;
+	net_buf_pull(pkt->frags, header_len);
+
+	r = zoap_packet_parse(&response, pkt);
+	if (r < 0) {
+		SYS_LOG_ERR("Invalid data received (err:%d)", r);
+		goto cleanup;
+	}
+
+	token = zoap_header_get_token(&response, &tkl);
+	pending = zoap_pending_received(&response, pendings,
+					NUM_PENDINGS);
+	if (pending) {
+		/* If necessary cancel retransmissions */
+	}
+
+	SYS_LOG_DBG("checking for reply from [%s]",
+		    sprint_ip_addr(&from_addr));
+	reply = zoap_response_received(&response, &from_addr,
+				       replies, NUM_REPLIES);
+	if (!reply) {
+		SYS_LOG_ERR("No handler for response");
+	} else {
+		SYS_LOG_DBG("reply handled reply:%p", reply);
+		zoap_reply_clear(reply);
+	}
+
+cleanup:
+	if (pkt) {
+		net_pkt_unref(pkt);
+	}
+}
+
+static void retransmit_request(struct k_work *work)
+{
+	struct zoap_pending *pending;
+	int r;
+
+	pending = zoap_pending_next_to_expire(pendings, NUM_PENDINGS);
+	if (!pending) {
+		return;
+	}
+
+	r = net_context_sendto(pending->pkt, &pending->addr,
+			       NET_SOCKADDR_MAX_SIZE,
+			       NULL, K_NO_WAIT, NULL, NULL);
+	if (r < 0) {
+		return;
+	}
+
+	if (!zoap_pending_cycle(pending)) {
+		zoap_pending_clear(pending);
+		return;
+	}
+
+	k_delayed_work_submit(&retransmit_work, pending->timeout);
+}
+
+static int transfer_request(struct zoap_block_context *ctx,
+			    const u8_t *token, u8_t tkl,
+			    zoap_reply_t reply_cb)
+{
+	struct zoap_packet request;
+	struct net_pkt *pkt = NULL;
+	struct zoap_pending *pending;
+	struct net_buf *frag;
+	struct zoap_reply *reply = NULL;
+	int ret;
+
+	/* send request */
+	pkt = net_pkt_get_tx(firmware_net_ctx, K_FOREVER);
```

Comment:
It might be better to use a timeout here as using K_FOREVER can be a bit problematic. You are checking the return value so using K_FOREVER does not make sense either.